vmm/src/device_manager.rs

```diff
@@ -4127,6 +4129,27 @@ impl Pausable for DeviceManager {
                 migratable.lock().unwrap().pause()?;
             }
         }
+        // On AArch64, the pause of device manager needs to trigger
+        // a "pause" of GIC, which will flush the GIC pending tables
+        // and ITS tables to guest RAM.
+        #[cfg(target_arch = "aarch64")]
+        {
+            let gic_device = Arc::clone(
+                self.get_interrupt_controller()
+                    .unwrap()
+                    .lock()
+                    .unwrap()
+                    .get_gic_device()
+                    .unwrap(),
+            );
+            gic_device
+                .lock()
+                .unwrap()
+                .as_any_concrete_mut()
+                .downcast_mut::<KvmGicV3Its>()
```

Comment:
Here you assume the `GicDevice` is gonna be a `KvmGicV3Its`, but you can't be sure about it. Since `GicDevice` is a trait, the proper way to handle this downcast is by trying and only if the downcasting succeeded we can use the object to call `pause()`, otherwise, just do nothing.

Reply:
Hi @sboeuf, I think for current AArch64 code, here (https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/vmm/src/vm.rs#L1088) we create a GIC device, and we are sure it is the KvmGicV3Its because of the definition of `create_gic()` (https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/arch/src/aarch64/gic/mod.rs#L202). We then pass it to device manager in here (https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/vmm/src/vm.rs#L1117). So if we get the GIC device here by `get_gic_device()` and do a downcast, we are sure this is a `KvmGicV3Its`.

But yes I agree the code is quite confusing. I will try to improve it. Thanks for the review :))

Reply:
Tested locally as CI is currently down. The logs are attached below
```
running 1 test
OK 1073741824
OK 1073741824
test tests::parallel::test_snapshot_restore has been running for over 60 seconds
test tests::parallel::test_snapshot_restore ... ok
```